### ISSUE TITLE
Web Server Gives Exception on An Already-Submitted Answer Even If Correct

### ISSUE THREAD

**Issue by zaratec** [picoCTF/picoCTF-web#46]
Just an odd sequence of events led to the following:

1. Tested a challenge on the local web server
2. Realized the challenge's challenge.py flag-generation was incorrect when I tried entering in the correct key
3. Fixed the challenge's challenge.py, packaged, reinstalled, undeployed the old, and deployed the new
4. Went to web server to re-enter the correct flag now that the challenge.py file was generating the correct flag
5. Got a "You or one of your teammates has already tried this solution" web exception even though it was the correct flag

This could potentially cause an issue if a challenge's flag generation breaks, teams enter in the correct flag, and we have to redeploy a corrected version of the challenge. For any team that had entered in the correct password, they would get the web exception, making them unable to get credit for their correct answer.

The issue stems from Ln 415-418 of https://github.com/picoCTF/picoCTF-web/blob/3360207442cb622658ff9c9b92aebeba4f812ee1/api/problem.py.

Since it is checked above at Ln 386 that the player is not submitting an answer to a problem that they've already solved, wouldn't it make more sense to put it inside checking the submission is correct (ie. if you hadn't submitted a correct answer yet, then check if it's correct, then see if you've already answered the problem). Here's a potential solution:

```
if submission["correct"]:
   api.cache.invalidate_memoization(api.stats.get_score, {"kwargs.tid":tid}, {"kwargs.uid":uid})
   api.cache.invalidate_memoization(get_unlocked_pids, {"args":tid})
   api.cache.invalidate_memoization(get_solved_problems, {"kwargs.tid":tid}, {"kwargs.uid":uid})
   api.cache.invalidate_memoization(api.stats.get_score_progression, {"kwargs.tid":tid}, {"kwargs.uid":uid})
   api.achievement.process_achievements("submit", {"uid": uid, "tid": tid, "pid": pid})
else:
   if (key, pid) in [(submission["key"], submission["pid"]) for submission in get_submissions(tid=tid)]:
      exp = WebException("You or one of your teammates has already tried this solution.")
      exp.data = {'code': 'repeat'}
      raise exp

db.submissions.insert(submission)
```

**Comment by cganas** [comment]
Yeah, I can see how this feels awkward. The one thing worth noting is that there are consistency issues regardless of order when the flag for a challenge is changed on the fly, e.g. my once correct submission is now incorrect because of an updated problem instance. The established way of dealing with this is by using any of these [calls](https://github.com/picoCTF/picoCTF-web/blob/master/api/problem.py#L504-L589) which will reevaluate the correctness of flags on a per problem basis. 

These calls are not currently exposed by the web interface but could find good home in the admin management interface. It's not obvious they exist (#45).
